mlir/lib/Dialect/Rock/Transforms/AccelEmitter.cpp

```diff
@@ -209,19 +226,22 @@ ArrayAttr MfmaEmitter::computeOutputTransforms(PatternRewriter &b, Location loc,
   int64_t inputSpansPerMfmaIn = mfmaAttr.inputSpansPerMfmaIn;
   int64_t blocksInOutRegs = mfmaAttr.blocksInOutRegs;
   int64_t blocksPerRepeat = (mPerRepeat * nPerRepeat) / (m * n);
-  int64_t wavesInKernelBlock = blockSize / waveSize;
 
   int64_t retNumElements = accVectorType.getNumElements();
   int64_t numElements = retNumElements * mRepeats * nRepeats * nResultVectors;
-  TopDownTMBuilder splitMemoryCoords(b, {"bid", "tid", "item"},
-                                     {gridSize, blockSize, numElements}, loc);
-
-  splitMemoryCoords.merge({"g", "m", "n"}, {0, 1, 2}, {"bid"},
-                          {gridSize / gStride, gStride / nBlocks, nBlocks});
-
-  splitMemoryCoords.merge(
-      {"wave", "m_tid", "n_tid"}, {3, 4, 5}, "tid",
-      {wavesInKernelBlock, waveSize / inputSpanLen, inputSpanLen});
+  TopDownTMBuilder splitMemoryCoords =
+      createTopSplitTMBuilder(b, loc, numElements, gridSize, blockSize);
+  if (gridSize.has_value()) {
+    splitMemoryCoords.merge(
+        {"g", "m", "n"}, {0, 1, 2}, {"bid"},
+        {gridSize.value() / gStride, gStride / nBlocks, nBlocks});
+  }
+  if (blockSize.has_value()) {
+    int64_t wavesInKernelBlock = blockSize.value() / waveSize;
+    splitMemoryCoords.merge(
+        {"wave", "m_tid", "n_tid"}, {3, 4, 5}, "tid",
```

Comment:
Let me be clear: the problem I'm seeing is that the indices `{3, 4, 5}` should not be `{3, 4, 5}`, they should be `{numGridDims, numGridDims + 1, numGridDims + 2}`, where `numGridDims == gridSize.has_value() ? 3 : 0`

Reply:
ah right got it..